sdk/core/System.ClientModel/src/Pipeline/ClientPipeline.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Primitives;
+
+public sealed partial class ClientPipeline
+{
+    private readonly int _perCallIndex;
+    private readonly int _perTryIndex;
+    private readonly int _beforeTransportIndex;
+
+    private readonly ReadOnlyMemory<PipelinePolicy> _policies;
+    private readonly PipelineTransport _transport;
+
+    private ClientPipeline(ReadOnlyMemory<PipelinePolicy> policies, int perCallIndex, int perTryIndex, int beforeTransportIndex)
+    {
+        if (perCallIndex > perTryIndex) throw new ArgumentOutOfRangeException(nameof(perCallIndex), "perCallIndex cannot be greater than perTryIndex.");
+
+        if (policies.Span[policies.Length - 1] is not PipelineTransport)
+        {
+            throw new ArgumentException("Last policy in the array must be of type 'PipelineTransport'.", nameof(policies));
```

Comment:
Same comment here about making this a Debug.Assert or even just validating this invariant via tests.

Reply:
This one I'll keep because we've been discussing a constructor/factory method that takes only an array of policies.  I don't know if we'll end up with it, but I'd be afraid of forgetting to add the validation back.  If you think there is a strong negative to this approach, @JoshLove-msft, please let me know so I can consider that in the mix - if needed I could file an issue to track adding the validation if we add the end-user API.